Assets/Scripts/Pathfinding/PathfindingPriorityQueue.cs

```diff
@@ -0,0 +1,134 @@
+﻿using UnityEngine;
+using System.Collections.Generic;
+using Priority_Queue;
+using System;
+
+/// <summary>
+/// Wraps the FastPriorityQueue class so that it's both easy-to-use,
+/// and faster than SimplePriorityQueue (which sports an O(n) Contains
+/// and an O(n) UpdatePriority -- not exactly ideal).
+/// </summary>
+public class PathfindingPriorityQueue<T>
+{
+    /// <summary>
+    /// The underlying FastPriorityQueue instance.
+    /// </summary>
+    protected FastPriorityQueue<WrappedNode> _underlyingQueue;
+
+    /// <summary>
+    /// The map between data and WrappedNodes.
+    /// Used to make operations like Contains and UpdatePriority more efficient.
+    /// </summary>
+    protected Dictionary<T, WrappedNode> _mapDataToWrappedNode;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PathfindingPriorityQueue`1"/> class.
+    /// </summary>
+    /// <param name="startingSize">The starting size.</param>
+    public PathfindingPriorityQueue(int startingSize = 10)
+    {
+        _underlyingQueue = new FastPriorityQueue<WrappedNode>(startingSize);
+        _mapDataToWrappedNode = new Dictionary<T, WrappedNode>();
+    }
+
+    /// <summary>
+    /// A version of a PriorityQueueNode that contains a reference to data.
+    /// </summary>
+    protected class WrappedNode : FastPriorityQueueNode
+    {
+        /// <summary>
+        /// The data that this WrappedNode represents in the queue.
+        /// </summary>
+        public readonly T data;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PathfindingPriorityQueue`1+WrappedNode"/> class.
+        /// </summary>
+        /// <param name="data">The data that this WrappedNode represents in the queue.</param>
+        public WrappedNode(T data)
+        {
+            this.data = data;
+        }
+    }
+
+    /// <summary>
+    /// Checks whether the PQ contains the specified data.
+    /// Uses a Dictionary for lookup, so it should only take O(1).
+    /// </summary>
+    /// <param name="data">Data.</param>
+    public bool Contains(T data)
+    {
+        return _mapDataToWrappedNode.ContainsKey(data);
+    }
+
+    /// <summary>
+    /// Enqueue the specified data and priority.
+    /// If the data already exists in the queue, it updates the priority instead.
+    /// Should take O(log n) -- O(1) amortized for the resizing, and O(log n) for the insertion.
+    /// </summary>
+    /// <param name="data">The data to be enqueued.</param>
+    /// <param name="priority">The priority of the data.</param>
+    public void Enqueue(T data, float priority)
+    {
+        if (_mapDataToWrappedNode.ContainsKey(data))
+        {
+            throw new InvalidOperationException("Can't re-enqueue a node that's already enqueued.");
```

Comment:
No, we obviously want the program to stop immediately so it isn't missed. Don't you agree? ;)